runtime/tr.source/trj9/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -5381,6 +5381,257 @@ bool genInstanceOfOrCheckCastNullTest(TR::Node* node, TR::CodeGenerator* cg, TR:
+ *     Function switches to outlined section and generate a JIT helper dispatch sequence.
+ *     It is responsibility of the caller to switch back to main line.
+ *
+ *  \param node
+ *     The instanceof node
+ *
+ *  \param cg
+ *     The code generator used to generate the instructions.
+ *
+ *  \param **deps
+ *     Place holder for the register dependency conditions which will e filled by buildDirectDispatch to
+ *     attach it to merge label of internal control flow.
+ *
+ *  \return
+ *     <c>TR::Register*</c> Return value register from helper call.
+ */
+static
+TR::Register* genInstanceOfOrCheckCastHelperCall(TR::Node *node, TR::CodeGenerator *cg, TR::Compilation *comp, TR::RegisterDependencyConditions **deps, TR::Register *resultReg, TR::LabelSymbol *callLabel, TR::LabelSymbol *doneLabel, TR_S390OutOfLineCodeSection **outlinedSlowPath)
+   {
+   if (!comp->getOption(TR_DisableOOL))
+      {
+      if (*outlinedSlowPath == NULL)
+         *outlinedSlowPath = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(callLabel, doneLabel, cg);
+      cg->getS390OutOfLineCodeSectionList().push_front(*outlinedSlowPath);
+      (*outlinedSlowPath)->swapInstructionListsWithCompilation();
+      }
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, callLabel);
+   // TODO : We can make this function a generic helper call function. Debug counter name then should be changed as per node.
+   cg->generateDebugCounter(TR::DebugCounter::debugCounterName(comp, "instanceOf/(%s)/Helper", comp->signature()),1,TR::DebugCounter::Undetermined);
+   TR::S390CHelperLinkage *helperLink =  static_cast<TR::S390CHelperLinkage*>(cg->createLinkage(TR_CHelper));
+   return helperLink->buildDirectDispatch(node, deps, resultReg);
+   } 
+
+
+/** \brief
+ *     Generates a dynamicCache test with helper call for instanceOf/ifInstanceOf node
+ *
+ *  \details
+ *     This funcition generates a sequence to check per site cache for object class and cast class before calling out to jitInstanceOf helper
+ */
+static
+void genInstanceOfDynamicCacheAndHelperCall(TR::Node *node, TR::CodeGenerator *cg, TR::Register *castClassReg, TR::Register *objClassReg, TR::Register *resultReg, TR_S390ScratchRegisterManager *srm, TR::LabelSymbol *doneLabel, TR::LabelSymbol *helperCallLabel, TR::LabelSymbol *dynamicCacheTestLabel, TR::LabelSymbol *branchLabel, TR::LabelSymbol *trueLabel, TR::LabelSymbol *falseLabel, bool dynamicCastClass, bool generateDynamicCache, bool cacheCastClass, bool ifInstanceOf, bool trueFallThrough )
+   {
+   TR_S390OutOfLineCodeSection *outlinedSlowPath = NULL;
+   TR::Compilation                *comp = cg->comp();
+   bool needResult = resultReg != NULL;
+   if (!castClassReg)
+      castClassReg = cg->evaluate(node->getSecondChild());
+   int32_t maxOnsiteCacheSlots = comp->getOptions()->getMaxOnsiteCacheSlotForInstanceOf();
+   TR::InstOpCode::Mnemonic loadHalfWordImmOp = TR::Compiler->target.is64Bit() ? TR::InstOpCode::LGHI : TR::InstOpCode::LHI;
```

Comment:
We should use an existing API here:
https://github.com/eclipse/omr/blob/master/compiler/z/codegen/InstOpCode.cpp#L421-L422